src/DataStructures/DynamicBuffer.hpp

```diff
@@ -0,0 +1,151 @@
+      buffer_.resize(number_of_vectors * number_of_grid_points);
+      set_references();
+    } else {
+      static_assert(
+          std::is_fundamental_v<T>,
+          "T was found to be neither a DataVector nor a fundamental type. "
+          "`DynamicBuffer` is not implemented yet for complex numbers.");
+      if (number_of_grid_points != 1) {
+        ERROR(
+            "DynamicBuffer must have number_of_grid_points == 1 when T is a "
+            "fundamental type but has number_of_grid_points = "
+            << number_of_grid_points_);
+      }
+    }
+  }
+
+  ~DynamicBuffer() = default;
+  DynamicBuffer(DynamicBuffer&& other) noexcept = default;
+  DynamicBuffer& operator=(DynamicBuffer&& other) = default;
+
+  DynamicBuffer(const DynamicBuffer& other) noexcept
+      : number_of_grid_points_(other.number_of_grid_points_) {
+    if constexpr (is_data_vector_type) {
+      data_.resize(other.size());
+      buffer_ = other.buffer_;
+      set_references();
+    } else {
+      data_ = other.data_;
+    }
+  }
+
+  DynamicBuffer& operator=(const DynamicBuffer& other) noexcept {
+    if (this == &other) {
+      return *this;
+    }
+    number_of_grid_points_ = other.number_of_grid_points_;
+    if constexpr (is_data_vector_type) {
+      data_.resize(other.size());
+      buffer_ = other.buffer_;
+      set_references();
+    } else {
+      data_ = other.data_;
+    }
+    return *this;
+  }
+
+  T& operator[](size_t index) { return data_[index]; }
+  T& at(size_t index) { return data_.at(index); }
+  const T& operator[](size_t index) const { return data_[index]; }
+  const T& at(size_t index) const { return data_.at(index); }
+
+  auto begin() { return data_.begin(); }
+  auto end() { return data_.end(); }
+  auto begin() const { return data_.begin(); }
+  auto end() const { return data_.end(); }
+
+  size_t size() const noexcept { return data_.size(); }
+
+  /// \cond
+  // clang-tidy: runtime-references
+  void pup(PUP::er& p) noexcept {
+    p | number_of_grid_points_;
+    p | buffer_;
+    if constexpr (is_data_vector_type) {
+      if (p.isUnpacking()) {
+        set_references();
+      }
+    } else {
+      p | data_;
+    }
+  }
+  /// \endcond
+
+ private:
+  // sets data references for all `data_` into `buffer_`
+  void set_references() {
+    for (size_t i = 0; i < size(); ++i) {
+      data_[i].set_data_ref(&buffer_[number_of_grid_points_ * i],
+                            number_of_grid_points_);
+    }
+  }
+
+  size_t number_of_grid_points_;
+  // vector of non-owning DataVectors pointing into `buffer_`. In case of
+  // fundamental type T the data is saved in `data_` directly.
+  std::vector<T> data_;
+  // memory buffer for all DataVectors. Unused in case of fundamental type T.
+  std::vector<double> buffer_;
+};
+
+template <typename T>
+bool operator==(const DynamicBuffer<T>& lhs,
+                const DynamicBuffer<T>& rhs) noexcept {
+  return std::equal(lhs.begin(), lhs.end(), rhs.begin());
```

Comment:
can you check the `number_of_grid_points_` too? Part of that check is to make sure the `pup` methods are correct. Then you can also just do `lhs.data_ == rhs.data_`. You'll need to make this a friend function, though